File: lc237_delNodeList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteNode(self, node):
        """
        :type node: ListNode
        :rtype: void Do not return anything, modify node in-place instead.
        """
        val=node.val #val to delete.
        #approach: move all values before, free the last one.
        #atleast two element, we can move always. Tricky part is to terminate the modified list before
        #its pointer reaches to last node.
        t=node
        while t:
            t.val = t.next.val
            if t.next.next == None:
                t.next=None
                break
            t=t.next
        
        return

File: lc237_delNodeList.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
void deleteNode(struct ListNode* node) {
    struct ListNode *t;
    t=node;
        while(t){  //4,5,6 delete 5, t points index1
            t->val=t->next->val;   //4,6,6 t points index1
            
            if(t->next->next == NULL){//we moved all vals we are pointing to the last node, however we should terminate the list
                // that means, we need access of previous node
                free(t->next->next);
                t->next=NULL;
                break;
            }
            t=t->next;
        }
    return;
}
